File: packages/python/peru_spatial_utils/ubigeo/search_districts.py
from typing import List

from ._normalize import normalize
from ._records import UbigeoRecord, districts


def _rank(name: str, query: str) -> int:
    if name == query:
        return 0
    if name.startswith(query):
        return 1
    if (" " + query) in name:
        return 2
    if query in name:
        return 3
    return -1


def search_districts(query: str, limit: int = 10) -> List[UbigeoRecord]:
    normalized = normalize(query)
    if not normalized:
        return []
    scored: List[tuple[int, int, UbigeoRecord]] = []
    for order, record in enumerate(districts().values()):
        rank = _rank(normalize(record["name"]), normalized)
        if rank >= 0:
            scored.append((rank, order, record))
    scored.sort(key=lambda item: (item[0], item[1]))
    return [record for _, _, record in scored[:limit]]

File: packages/dart/lib/src/ubigeo/search_districts.dart
import 'normalize_text.dart';
import 'ubigeo_record.dart';
import 'ubigeo_store.dart';

List<UbigeoRecord> searchDistricts(String query, {int limit = 10}) {
  final normalizedQuery = normalizeText(query);
  if (normalizedQuery.isEmpty) return [];

  int rank(String name) {
    final n = normalizeText(name);
    if (n == normalizedQuery) return 0;
    if (n.startsWith(normalizedQuery)) return 1;
    if (RegExp(r'\b' + RegExp.escape(normalizedQuery)).hasMatch(n)) return 2;
    if (n.contains(normalizedQuery)) return 3;
    return -1;
  }

  final matches = <MapEntry<int, UbigeoRecord>>[];
  var order = 0;
  for (final record in UbigeoStore.districts.values) {
    final r = rank(record.name);
    if (r >= 0) {
      matches.add(MapEntry(r * 1000000 + order, record));
    }
    order++;
  }

  matches.sort((a, b) => a.key.compareTo(b.key));

  return matches.take(limit).map((e) => e.value).toList();
}
